evidence/pool.go

```diff
@@ -424,25 +263,30 @@ func (evpool *Pool) fastCheck(ev types.Evidence) bool {
 			evpool.logger.Error("Failed to load evidence", "err", err, "evidence", lcae)
 			return false
 		}
-		evInfo, err := bytesToInfo(evBytes)
+		var trustedPb tmproto.LightClientAttackEvidence
+		err = trustedPb.Unmarshal(evBytes)
 		if err != nil {
-			evpool.logger.Error("Failed to convert evidence from proto", "err", err, "evidence", lcae)
+			evpool.logger.Error("Failed to convert evidence from bytes", "err", err, "evidence", lcae)
 			return false
 		}
-		// ensure that all the validators that the evidence pool have found to be malicious
-		// are present in the list of commit signatures in the conflicting block
-	OUTER:
-		for _, sig := range lcae.ConflictingBlock.Commit.Signatures {
-			for _, val := range evInfo.Validators {
-				if bytes.Equal(val.Address, sig.ValidatorAddress) {
-					continue OUTER
-				}
-			}
-			// a validator we know is malicious is not included in the commit
-			evpool.logger.Info("Fast check failed: a validator we know is malicious is not " +
-				"in the commit sigs. Reverting to full verification")
+		trustedEv, err := types.LightClientAttackEvidenceFromProto(&trustedPb)
+		if err != nil {
+			evpool.logger.Error("Failed to convert evidence from protobuf", "err", err, "evidence", lcae)
+			return false
+		}
+		// ensure that all the validators that the evidence pool has matches the validators
+		// in this evidence
+		if len(trustedEv.ByzantineValidators) != len(lcae.ByzantineValidators) {
 			return false
 		}
+		for idx, val := range trustedEv.ByzantineValidators {
```

Comment:
This assumes that both evidence lists use the same ordering for validators. Will this always be true? We may at some point want to change the ordering of validators (like we did when we changed to order by power instead of ID), but old blocks with the previous ordering should ideally still work even though the new code uses a new order. To handle this, we may want to consider sorting these before comparing, or build a map of addresses that we compare instead.

Reply:
presorting sounds fine by me

Reply:
If we keep the previous ordering method for validator sets of older blocks then this should not be a problem because the evidence pool will be getting the validator set at the same height and thus they should be sorted in the same way. This only is a problem when some nodes update the ordering of older validator sets are other nodes (who perhaps haven't upgraded yet) don't. Then we could potentially get different ordering.

Reply:
I guess the main question is whether we really care about the order here. If we don't, then we shouldn't implicitly rely on it.

Reply:
Yeah that does makes sense. I was purely relying on it as just a performance optimization

Reply:
I don't think it should be possible for this to ever happen but I have sorted it nonetheless